WebDriverAgentLib/Categories/XCUIElement+FBClassChain.m

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "XCUIElement+FBClassChain.h"
+#import "FBClassChainQueryParser.h"
+#import "FBElementUtils.h"
+#import "XCElementSnapshot.h"
+#import "XCUIElement+FBUtilities.h"
+
+NSString *const FBClassChainQueryParseException = @"FBClassChainQueryParseException";
+
+@implementation XCUIElement (FBClassChain)
+
+- (NSArray<XCUIElement *> *)fb_descendantsMatchingClassChain:(NSString *)classChainQuery shouldReturnAfterFirstMatch:(BOOL)shouldReturnAfterFirstMatch
+{
+  NSError *error;
+  FBClassChain parsedChain = [FBClassChainQueryParser parseQuery:classChainQuery error:&error];
+  if (nil == parsedChain) {
+    @throw [NSException exceptionWithName:FBClassChainQueryParseException reason:error.localizedDescription userInfo:error.userInfo];
+    return nil;
+  }
+  NSMutableArray *unmatchedSnapshots = [NSMutableArray arrayWithArray:[self fb_lookupChain:parsedChain root:self.fb_lastSnapshot]];
+  if (0 == unmatchedSnapshots.count) {
+    return @[];
+  }
+  if (shouldReturnAfterFirstMatch) {
+    [unmatchedSnapshots removeObjectsInRange:NSMakeRange(1, unmatchedSnapshots.count - 1)];
+  }
+  XCUIElementQuery *elementsQuery = [self childrenMatchingType:[parsedChain firstObject].type];
+  for (NSUInteger level = 1; level < parsedChain.count; ++level) {
+    elementsQuery = [elementsQuery childrenMatchingType:[parsedChain objectAtIndex:level].type];
+  }
+  NSMutableArray *result = [NSMutableArray array];
+  NSMutableArray *candidateElements = [NSMutableArray arrayWithArray:elementsQuery.allElementsBoundByIndex];
+  if ([parsedChain lastObject].position < 0 && candidateElements.count > 1) {
+    candidateElements = [[[candidateElements reverseObjectEnumerator] allObjects] mutableCopy];
+  }
+  NSUInteger candidateElementIdx = 0;
+  while (candidateElementIdx < candidateElements.count) {
```

Comment:
> [candidateElements removeObjectAtIndex:candidateElementIdx];

because of that. I cannot change the structure being currently iterated, otherwise this will corrupt my iterator.

I want to remove matched items immediately to reduce iterations count